internal/monitors/haproxy/helpers.go

```diff
@@ -0,0 +1,175 @@
+package haproxy
+
+import (
+	"bufio"
+	"crypto/tls"
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/signalfx/golib/datapoint"
+	logger "github.com/sirupsen/logrus"
+)
+
+func newStatPageDatapoints(body io.ReadCloser, sfxMetricNames map[string]string, proxiesToMonitor map[string]bool) []*datapoint.Datapoint {
+	dps := make([]*datapoint.Datapoint, 0)
+	for _, metricValuePairs := range statsPageMetricValuePairs(body) {
+		if len(proxiesToMonitor) != 0 && !proxiesToMonitor[metricValuePairs["pxname"]] && !proxiesToMonitor[metricValuePairs["svname"]] {
+			continue
+		}
+		for metric, value := range metricValuePairs {
+			if dp := newDatapoint(sfxMetricNames[metric], value); dp != nil {
+				dp.Dimensions["proxy_name"] = metricValuePairs["pxname"]
+				dp.Dimensions["service_name"] = metricValuePairs["svname"]
+				dp.Dimensions["process_num"] = metricValuePairs["pid"]
+				dps = append(dps, dp)
+			}
+		}
+	}
+	return dps
+}
+
+func newShowStatCommandDatapoints(body io.ReadCloser, sfxMetricNames map[string]string, proxiesToMonitor map[string]bool) []*datapoint.Datapoint {
+	return newStatPageDatapoints(body, sfxMetricNames, proxiesToMonitor)
+}
+
+func newShowInfoCommandDatapoints(body io.ReadCloser, sfxMetricNames map[string]string) []*datapoint.Datapoint {
+	dps := make([]*datapoint.Datapoint, 0)
+	for _, metricValuePairs := range showInfoCommandMetricValuePairs(body) {
+		for metric, value := range metricValuePairs {
+			if dp := newDatapoint(sfxMetricNames[metric], value); dp != nil {
+				dp.Dimensions["process_num"] = metricValuePairs["Process_num"]
+				dps = append(dps, dp)
+			}
+		}
+	}
+	return dps
+}
+
+func statsPageMetricValuePairs(body io.ReadCloser) map[int]map[string]string /*([]*datapoint.Datapoint, error)*/ {
+	defer closeBody(body)
+	r := csv.NewReader(body)
+	r.TrimLeadingSpace = true
+	r.TrailingComma = true
+	rows := map[int]map[string]string{}
+	if table, err := r.ReadAll(); err == nil && len(table) > 1 {
+		// fixing first column header because it is '# pxname' instead of 'pxname'
+		table[0][0] = "pxname"
+		for rowIndex := 1; rowIndex < len(table); rowIndex++ {
+			if rows[rowIndex-1] == nil {
+				rows[rowIndex-1] = map[string]string{}
+			}
+			for j, colName := range table[0] {
+				rows[rowIndex-1][colName] = table[rowIndex][j]
+			}
+		}
+	}
+	return rows
+}
+
+func showInfoCommandMetricValuePairs(body io.ReadCloser) map[int]map[string]string /*([]*datapoint.Datapoint, error)*/ {
+	defer closeBody(body)
+	sc := bufio.NewScanner(body)
+	rows := map[int]map[string]string{}
+	for sc.Scan() {
+		s := strings.SplitN(sc.Text(), ":", 2)
+		if len(s) != 2 {
+			logger.Debugf("could not split string '%s' into 2 substrings using separator ':'", sc.Text())
+			continue
+		}
+		if rows[0] == nil {
+			rows[0] = map[string]string{}
+		}
+		rows[0][strings.TrimSpace(s[0])] = strings.TrimSpace(s[1])
```

Comment:
So `rows` will always only have a `0` key and nothing else?

Reply:
Fixed this issue. There can only be one row.